### PR TITLE
feat(freespace_planning_algorithms): use distance to nearest obstacle to improve path planning

### PR DESCRIPTION
**Pull request by mkquda**
## Description
Current A-star planner does not use any information regarding distance to obstacle for path planning, it only checks for collision for each expanded node.
As a result generated path can navigate close to obstacles and cause the vehicle to stop frequently as it follows the path.

## Changes
- Includes changes from preceding [PR](https://github.com/autowarefoundation/autoware.universe/pull/8068)
- Compute an EDT (euclidean distance transform) map to store the distance to nearest obstacle for each grid cell
- Use EDT data to add cost to account for closeness to obstacles
- Use EDT value to minimize collision check necessity (by comparing EDT value to vehicle largest and smallest dimensions)
- Use distance to obstacle and distance to goal to vary node expansion distance (shorter primitives near obstacles / goal, longer primitives far from obstacles / goal)

## Related links

**Preceding PR's:**

- https://github.com/autowarefoundation/autoware.universe/pull/8068

## How was this PR tested?
tested on psim.
compared performance with unit tests: (see [here](https://github.com/autowarefoundation/autoware.universe/pull/8091) for complete comparison)

## Notes for reviewers

This PR includes changes from a preceding [PR](https://github.com/autowarefoundation/autoware.universe/pull/8068). So please review the PR's in order. 

#### Additions and removals

| Change type   | Parameter Name | Type     | Default Value | Description       |
|:--------------|:---------------|:---------|:--------------|:------------------|
| Added | `adapt_expansion_distance`  | `bool` | `true` | flag to enable / disable variable A-star expansion distance |
| Added | `obstacle_distance_weight`   | `double` | `1.5` | additional cost factor for closeness to obstacles |

## Effects on system behavior

When using free space planner, path should navigate away from obstacles when possible.

### CODE REVIEW

File: planning/autoware_freespace_planning_algorithms/include/autoware/freespace_planning_algorithms/abstract_algorithm.hpp [hunk 1]
```diff
@@ -84,6 +88,13 @@ struct VehicleShape
     max_steering(vehicle_info.max_steer_angle_rad),
     base2back(vehicle_info.rear_overhang_m + margin / 2.0)
   {
+    setMinMaxDimension();
+  }
+
+  void setMinMaxDimension()
+  {
+    min_dimension = std::min(0.5 * width, base2back);
+    max_dimension = std::hypot(length - base2back, 0.5 * width);
```
Comment (kosuke55) [inline planning/autoware_freespace_planning_algorithms/include/autoware/freespace_planning_algorithms/abstract_algorithm.hpp:97]: [nits]
this is rare but some vehicle may have longer base2back than base2front.

File: planning/autoware_freespace_planning_algorithms/src/astar_search.cpp [hunk 2]
```diff
@@ -342,10 +341,27 @@ void AstarSearch::setPath(const AstarNode & goal_node)
 
   geometry_msgs::msg::PoseStamped pose;
   pose.header = header;
+
+  const auto interpolate = [this, &pose](const AstarNode & node) {
+    if (node.parent == nullptr || !astar_param_.adapt_expansion_distance) return;
+    const auto parent_pose = node2pose(*node.parent);
+    const double distance_2d = calcDistance2d(node2pose(node), parent_pose);
+    int n = static_cast<int>(distance_2d / min_expansion_dist_);
```
Comment (kosuke55) [inline planning/autoware_freespace_planning_algorithms/src/astar_search.cpp:349]:
```suggestion
    const int n = static_cast<int>(distance_2d / min_expansion_dist_);
```

File: planning/autoware_freespace_planning_algorithms/src/astar_search.cpp [hunk 3]
```diff
@@ -342,10 +341,27 @@ void AstarSearch::setPath(const AstarNode & goal_node)
...
+    for (int i = 1; i < n; ++i) {
+      double dist = ((distance_2d * i) / n) * (node.is_back ? -1.0 : 1.0);
+      double steering = node.steering_index * steering_resolution_;
+      auto local_pose = kinematic_bicycle_model::getPose(
```
Comment (kosuke55) [inline planning/autoware_freespace_planning_algorithms/src/astar_search.cpp:353]:
```suggestion
      const double dist = ((distance_2d * i) / n) * (node.is_back ? -1.0 : 1.0);
      const double steering = node.steering_index * steering_resolution_;
      const auto local_pose = kinematic_bicycle_model::getPose(
```

File: planning/autoware_freespace_planning_algorithms/src/astar_search.cpp [hunk 4]
```diff
@@ -252,63 +232,76 @@ bool AstarSearch::search()
...
+    const auto next_pose = kinematic_bicycle_model::getPose(
+      current_pose, collision_vehicle_shape_.base_length, steering, distance);
     const auto next_index = pose2index(costmap_, next_pose, planner_common_param_.theta_size);
 
     if (isOutOfRange(next_index) || isObs(next_index)) continue;
 
     AstarNode * next_node = &graph_[getKey(next_index)];
     if (next_node->status == NodeStatus::Closed || detectCollision(next_index)) continue;
 
+    double distance_to_obs = getObstacleEDT(next_index);
```
Comment (kosuke55) [inline planning/autoware_freespace_planning_algorithms/src/astar_search.cpp:265]:
```suggestion
    const double distance_to_obs = getObstacleEDT(next_index);
```

**Review by maxime-clem — APPROVED** [review]
Tested the PR and the freespace planning looks good.

**Review by armaganarsln — DISMISSED** [review]
Can you please test with this issue: 
- https://github.com/autowarefoundation/autoware.universe/issues/8342

File: planning/autoware_freespace_planning_algorithms/include/autoware/freespace_planning_algorithms/abstract_algorithm.hpp [hunk 5]
```diff
@@ -156,6 +173,15 @@ class AbstractPlanningAlgorithm
     std::vector<IndexXY> & vertex_indexes_2d) const;
   bool detectBoundaryExit(const IndexXYT & base_index) const;
   bool detectCollision(const IndexXYT & base_index) const;
+  bool detectCollision(const geometry_msgs::msg::Pose & base_pose) const;
+
+  /// @brief Computes the euclidean distance to the nearest obstacle for each grid cell.
+  /// @cite T., Saito, and J., Toriwaki "New algorithms for euclidean distance transformation of an
```
Comment (maxime-clem) [inline planning/autoware_freespace_planning_algorithms/include/autoware/freespace_planning_algorithms/abstract_algorithm.hpp:179]:
```suggestion
  // cspell: ignore Toriwaki
  /// @cite T., Saito, and J., Toriwaki "New algorithms for euclidean distance transformation of an
```

**Comment by maxime-clem** [comment]
@armaganarsln If it is okay I will dismiss your request for changes so that we can move forward with improving the _freespace planner_.
The tests you requested show that the current _parking_ feature still has issues, but improvements to parking should be done in separate PRs.

**Review by kosuke55 — APPROVED** [review]

**Comment by armaganarsln** [comment]
> @armaganarsln If it is okay I will dismiss your request for changes so that we can move forward with improving the _freespace planner_. The tests you requested show that the current _parking_ feature still has issues, but improvements to parking should be done in separate PRs.

For me its ok as long as in the end the tests are made in simulation. Do you want me to delete my change requests on the other PRs as well?

**Pull request merged**